File: android/src/main/kotlin/media/bcc/bccm_player/PlaybackApiImpl.kt
package media.bcc.bccm_player

import android.content.Intent
import android.util.Log
import com.google.android.gms.cast.framework.CastButtonFactory
import io.flutter.embedding.android.FlutterFragmentActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import media.bcc.bccm_player.pigeon.PlaybackPlatformApi
import media.bcc.bccm_player.pigeon.PlaybackPlatformApi.BufferMode
import media.bcc.bccm_player.players.chromecast.CastExpandedControlsActivity
import media.bcc.bccm_player.players.chromecast.CastPlayerController
import media.bcc.bccm_player.utils.DevicePerformanceManager
import media.bcc.bccm_player.utils.toMedia3Type


class PlaybackApiImpl(private val plugin: BccmPlayerPlugin) :
    PlaybackPlatformApi.PlaybackPlatformPigeon {
    private val mainScope = CoroutineScope(Dispatchers.Main + Job())

    override fun attach(result: PlaybackPlatformApi.VoidResult) {
        Log.d("bccm", "attaching plugin")
        // Extremely important to call result.success or result.fail
        plugin.attach(onComplete = {
            result.success()
        })
    }

    override fun setNpawConfig(config: PlaybackPlatformApi.NpawConfig?) {
        Log.d("bccm", "PlaybackPigeon: Setting npawConfig")
        mainScope.launch {
            BccmPlayerPluginSingleton.npawConfigState.update { config }
        }
    }

    override fun setAppConfig(config: PlaybackPlatformApi.AppConfig?) {
        Log.d("bccm", "PlaybackPigeon: Setting appConfig")
        mainScope.launch {
            BccmPlayerPluginSingleton.appConfigState.update { config }
        }
    }

    override fun getTracks(
        playerId: String?,
        result: PlaybackPlatformApi.NullableResult<PlaybackPlatformApi.PlayerTracksSnapshot>
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController =
            if (playerId != null) playbackService.getController(playerId) else playbackService.getPrimaryController()
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }
        result.success(playerController.getTracksSnapshot())
    }

    override fun getPlayerState(
        playerId: String?,
        result: PlaybackPlatformApi.NullableResult<PlaybackPlatformApi.PlayerStateSnapshot>
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController =
            if (playerId != null) playbackService.getController(playerId) else playbackService.getPrimaryController()
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }
        result.success(playerController.getPlayerStateSnapshot())
    }

    override fun setSelectedTrack(
        playerId: String,
        type: PlaybackPlatformApi.TrackType,
        trackId: String?,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController = playbackService.getController(playerId)
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }
        playerController.setSelectedTrack(type.toMedia3Type(), trackId)
        result.success()
    }

    override fun setPlaybackSpeed(
        playerId: String,
        speed: Double,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController = playbackService.getController(playerId)
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }
        playerController.setPlaybackSpeed(speed.toFloat())
        result.success()
    }

    override fun newPlayer(
        bufferMode: BufferMode?,
        disableNpaw: Boolean?,
        result: PlaybackPlatformApi.Result<String>
    ) {
        Log.d("bccm", "PlaybackPigeon: newPlayer()")
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController = playbackService.newPlayer(
            bufferMode ?: BufferMode.STANDARD,
            disableNpaw ?: false
        )
        result.success(playerController.id)
    }

    override fun createVideoTexture(result: PlaybackPlatformApi.Result<Long>) {
        Log.d("bccm", "PlaybackPigeon: createVideoTexture()")
        val texture = plugin.createTexture()
        if (texture == null) {
            result.error(Error("Could not create texture"))
            return
        }
        result.success(texture.id())
    }

    override fun switchToVideoTexture(
        playerId: String,
        textureId: Long,
        result: PlaybackPlatformApi.Result<Long>
    ) {
        Log.d("bccm", "PlaybackPigeon: switchToVideoTexture()")
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController = playbackService.getController(playerId)
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }
        val texture = plugin.getTexture(textureId)
        if (texture == null) {
            result.error(Error("Texture with id $textureId does not exist."))
            return
        }
        playerController.setVideoTexture(texture)
        result.success(texture.id())
    }

    override fun disposeVideoTexture(
        textureId: Long,
        result: PlaybackPlatformApi.Result<Boolean>
    ): Unit {
        Log.d("bccm", "PlaybackPigeon: disposeVideoTexture()")
        result.success(plugin.releaseTexture(textureId))
    }

    override fun disposePlayer(playerId: String, result: PlaybackPlatformApi.Result<Boolean>) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error("Playback service doesnt exist"))
            return
        }
        val didDispose = playbackService.disposePlayer(playerId)
        result.success(didDispose);
    }

    override fun replaceCurrentMediaItem(
        playerId: String,
        mediaItem: PlaybackPlatformApi.MediaItem,
        playbackPositionFromPrimary: Boolean?,
        autoplay: Boolean?,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        if (playbackPositionFromPrimary == true) {
            mediaItem.playbackStartPositionMs =
                playbackService.getPrimaryController()?.player?.currentPosition?.toDouble()
        }

        val playerController = playbackService.getController(playerId)
        if (playerController == null) {
            result.error(Error("Player with id $playerId does not exist."))
            return
        }

        playerController.replaceCurrentMediaItem(mediaItem, autoplay)
        result.success()
    }

    override fun setPlayerViewVisibility(viewId: Long, visible: Boolean) {
        mainScope.launch {
            BccmPlayerPluginSingleton.eventBus.emit(SetPlayerViewVisibilityEvent(viewId, visible))
        }
    }

    override fun queueMediaItem(
        playerId: String,
        mediaItem: PlaybackPlatformApi.MediaItem,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")
        playerController.queueMediaItem(mediaItem)
        result.success()
    }

    override fun setPrimary(id: String, result: PlaybackPlatformApi.VoidResult) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }

        playbackService.setPrimary(id)
        result.success()
    }

    override fun play(playerId: String) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.play()
    }

    override fun seekTo(
        playerId: String,
        positionMs: Double,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")
        try {
            playerController.player.seekTo(positionMs.toLong());
            result.success();
        } catch (e: Exception) {
            result.error(e)
        }
    }

    override fun pause(playerId: String) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.pause()
    }

    override fun setRepeatMode(
        playerId: String,
        repeatMode: PlaybackPlatformApi.RepeatMode,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.setRepeatMode(repeatMode)
        result.success()
    }

    override fun setVolume(
        playerId: String,
        volume: Double,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: return result.error(Error("Player with id $playerId does not exist."))

        playerController.setVolume(volume)
        result.success()
    }

    override fun stop(playerId: String, reset: Boolean) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.stop(reset)
    }

    override fun exitFullscreen(playerId: String) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.currentPlayerViewController?.exitFullscreen()
    }

    override fun enterFullscreen(playerId: String) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")

        playerController.currentPlayerViewController?.enterFullscreen()
    }

    override fun setMixWithOthers(
        playerId: String,
        mixWithOthers: Boolean,
        result: PlaybackPlatformApi.VoidResult
    ) {
        val playbackService = plugin.getPlaybackService() ?: return
        val playerController = playbackService.getController(playerId)
            ?: throw Error("Player with id $playerId does not exist.")
        playerController.setMixWithOthers(mixWithOthers)
    }

    override fun getChromecastState(result: PlaybackPlatformApi.NullableResult<PlaybackPlatformApi.ChromecastState>) {
        val playbackService = plugin.getPlaybackService()
        if (playbackService == null) {
            result.error(Error())
            return
        }
        val cc = playbackService.getController("chromecast")
        if (cc == null || cc !is CastPlayerController) {
            return
        }
        result.success(cc.getState())
    }

    override fun openExpandedCastController() {
        val intent = Intent(
            BccmPlayerPluginSingleton.activityState.value,
            CastExpandedControlsActivity::class.java
        )
        BccmPlayerPluginSingleton.activityState.value?.startActivity(intent)
    }

    override fun openCastDialog() {
        val activity =
            (BccmPlayerPluginSingleton.activityState.value as? FlutterFragmentActivity) ?: return
        val fm =
            activity.supportFragmentManager
                ?: return
        val btn = androidx.mediarouter.app.MediaRouteButton(activity)
        CastButtonFactory.setUpMediaRouteButton(activity, btn)
        btn.onAttachedToWindow()
        btn.showDialog()
    }

    override fun fetchMediaInfo(
        url: String,
        mimeType: String?,
        result: PlaybackPlatformApi.Result<PlaybackPlatformApi.MediaInfo>
    ) {
        val context = BccmPlayerPluginSingleton.activityState.value;
        if (context != null) {
            mainScope.launch {
                try {
                    val mediaInfo = MediaInfoFetcher.fetchMediaInfo(context, url, mimeType)
                    result.success(mediaInfo)
                } catch (e: Exception) {
                    result.error(e)
                }
            }
        } else {
            result.error(Error("Not attached to activity"))
        }
    }

    override fun getAndroidPerformanceClass(result: PlaybackPlatformApi.Result<Long>) {
        val context = plugin.getApplicationContext()
        if (context == null) {
            result.error(Error("No context available"))
            return
        }
        var performanceClass =
            DevicePerformanceManager.getInstance(context)?.mediaPerformanceClass?.toLong()
        if (performanceClass == 0L) {
            performanceClass = null
        }
        if (performanceClass != null) {
            result.success(performanceClass)
        } else {
            result.error(Error("Could not get performance class"))
        }
    }
}

File: ios/Classes/PlaybackApiImpl.swift
import AVKit
import Foundation
import GoogleCast

// Implementation of the PlaybackPlatformPigeon
// See pigeons/playback_platform_pigeon.dart
// TODO: this file should be a pure api towards flutter,
// we should move the "players" array and state into a dedicated class
public class PlaybackApiImpl: NSObject, PlaybackPlatformPigeon {

    public func getAndroidPerformanceClass(completion: @escaping (NSNumber?, FlutterError?) -> Void) {
        completion(nil, nil)
    }

    public func switchToVideoTexture(forPlayer playerId: String, textureId: Int, completion: @escaping (NSNumber?, FlutterError?) -> Void) {
        completion(nil, FlutterError(code: "PlaybackApiImpl", message: "Textures are not implemented on iOS. Wrap with Platform.isAndroid().", details: nil))
    }

    public func createVideoTexture(_ completion: @escaping (NSNumber?, FlutterError?) -> Void) {
        completion(nil, FlutterError(code: "PlaybackApiImpl", message: "Textures are not implemented on iOS. Wrap with Platform.isAndroid().", details: nil))
    }

    public func disposeVideoTexture(_ textureId: Int, completion: @escaping (NSNumber?, FlutterError?) -> Void) {
        completion(nil, FlutterError(code: "PlaybackApiImpl", message: "Textures are not implemented on iOS. Wrap with Platform.isAndroid().", details: nil))
    }

    var players = [PlayerController]()
    private var primaryPlayerId: String? = nil
    private var previousPrimaryPlayerId: String? = nil
    let playbackListener: PlaybackListenerPigeon
    let chromecastPigeon: ChromecastPigeon
    let queueManagerPigeon: QueueManagerPigeon
    var npawConfig: NpawConfig? = nil
    var appConfig: AppConfig? = nil

    init(chromecastPigeon: ChromecastPigeon, playbackListener: PlaybackListenerPigeon, queueManagerPigeon: QueueManagerPigeon) {
        self.playbackListener = playbackListener
        self.chromecastPigeon = chromecastPigeon
        self.queueManagerPigeon = queueManagerPigeon
        super.init()
        let castPlayerController = CastPlayerController(playbackApi: self)
        players.append(castPlayerController)
        newPlayer(nil, disableNpaw: false, completion: { playerId, _ in
            if playerId != nil {
                self.setPrimary(playerId!, completion: { _ in })
            }
        })
    }

    public func attach(completion: @escaping (FlutterError?) -> Void) {
        // stop and remove all of type AVQueuePlayerController except primary
        for p in players {
            if p is AVQueuePlayerController && p.id != primaryPlayerId {
                p.stop(reset: true)
            }
            let event = PlayerStateUpdateEvent.make(withPlayerId: p.id, snapshot: p.getPlayerStateSnapshot())
            playbackListener.onPlayerStateUpdate(event, completion: { _ in })
        }
        players.removeAll(where: { $0 is AVQueuePlayerController && $0.id != primaryPlayerId })

        completion(nil)
    }

    public func setAppConfig(_ config: AppConfig?, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        appConfig = config
        for p in players {
            p.updateAppConfig(appConfig: appConfig)
        }
    }

    public func setPlayerViewVisibility(_ viewId: Int, visible: Bool, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {}

    public func setNpawConfig(_ config: NpawConfig?, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        npawConfig = config
        for p in players {
            p.setNpawConfig(npawConfig: npawConfig)
        }
    }

    public func getPlayer(_ id: String) -> PlayerController? {
        players.first(where: { $0.id == id })
    }

    public func getPrimaryPlayer() -> PlayerController? {
        players.first(where: { $0.id == primaryPlayerId })
    }

    public func setPrimary(_ id: String, completion: @escaping (FlutterError?) -> Void) {
        if primaryPlayerId == id { return }
        previousPrimaryPlayerId = primaryPlayerId
        primaryPlayerId = id
        getPrimaryPlayer()?.hasBecomePrimary()
        playbackListener.onPrimaryPlayerChanged(PrimaryPlayerChangedEvent.make(withPlayerId: id), completion: { _ in })
        completion(nil)
    }

    public func newPlayer(_ bufferModeBoxed: BufferModeBox?, disableNpaw: NSNumber?, completion: @escaping (String?, FlutterError?) -> Void) {
        let bufferMode = bufferModeBoxed?.value ?? BufferMode.standard
        let player = AVQueuePlayerController(
            playbackListener: playbackListener,
            bufferMode: bufferMode,
            npawConfig: npawConfig,
            appConfig: appConfig,
            disableNpaw: disableNpaw?.boolValue,
            queueManagerPigeon: queueManagerPigeon
        )
        players.append(player)

        completion(player.id, nil)
    }

    public func disposePlayer(_ playerId: String, completion: @escaping (NSNumber?, FlutterError?) -> Void) {
        guard let player = getPlayer(playerId) else {
            completion(false, nil)
            return
        }
        player.stop(reset: true)
        players.removeAll(where: { $0.id == playerId })
        completion(true, nil)
    }

    public func getChromecastState(_ completion: @escaping (ChromecastState?, FlutterError?) -> Void) {
        let castPlayer = players.first(where: { $0.id == CastPlayerController.DEFAULT_ID })
        let mediaItem = castPlayer?.getCurrentItem()

        let connectionStateRaw = GCKCastContext.sharedInstance().castState.rawValue + 1
        let connectionState = CastConnectionState(rawValue: UInt(connectionStateRaw))
        if connectionState != nil {
            completion(ChromecastState.make(with: connectionState!, mediaItem: mediaItem), nil)
        } else {
            completion(ChromecastState.make(with: CastConnectionState.noDevicesAvailable, mediaItem: mediaItem), nil)
        }
    }

    public func openExpandedCastController(_ error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        GCKCastContext.sharedInstance().presentDefaultExpandedMediaControls()
    }

    public func openCastDialog(_ error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        GCKCastContext.sharedInstance().presentCastDialog()
    }

    public func replaceCurrentMediaItem(_ playerId: String, mediaItem: MediaItem, playbackPositionFromPrimary: NSNumber?, autoplay: NSNumber?, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)

        updateAudioSession()
        player?.replaceCurrentMediaItem(mediaItem, autoplay: autoplay, completion: completion)
    }

    public func getTracks(_ playerId: String?, completion: @escaping (PlayerTracksSnapshot?, FlutterError?) -> Void) {
        let player = playerId == nil ? getPrimaryPlayer() : getPlayer(playerId!)
        let snapshot = player?.getPlayerTracksSnapshot()
        completion(snapshot, nil)
    }

    public func getPlayerState(_ playerId: String?, completion: @escaping (PlayerStateSnapshot?, FlutterError?) -> Void) {
        let player = playerId == nil ? getPrimaryPlayer() : getPlayer(playerId!)
        let snapshot = player?.getPlayerStateSnapshot()
        completion(snapshot, nil)
    }

    public func setSelectedTrack(_ playerId: String, type: TrackType, trackId: String?, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)
        player?.setSelectedTrack(type: type, trackId: trackId)
        completion(nil)
    }

    public func setPlaybackSpeed(_ playerId: String, speed: Double, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)
        player?.setPlaybackSpeed(Float(speed))
        completion(nil)
    }

    public func setVolume(_ playerId: String, volume: Double, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)
        player?.setVolume(Float(volume))
        completion(nil)
    }

    public func setRepeatMode(_ playerId: String, repeatMode: RepeatMode, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)
        let repeatMode = repeatMode
        player?.setRepeatMode(repeatMode)
        completion(nil)
    }

    public func play(_ playerId: String, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        let player = getPlayer(playerId)
        player?.play()
    }

    public func seek(_ playerId: String, positionMs: Double, completion: @escaping (FlutterError?) -> Void) {
        let player = getPlayer(playerId)
        player?.seekTo(Int64(positionMs)) { _ in
            completion(nil)
        }
    }

    public func pause(_ playerId: String, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        let player = getPlayer(playerId)
        player?.pause()
    }

    public func stop(_ playerId: String, reset: Bool, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        let player = getPlayer(playerId)
        player?.stop(reset: reset)
    }

    public func exitFullscreen(_ playerId: String, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        let player = getPlayer(playerId)
        player?.exitFullscreen()
    }

    public func enterFullscreen(_ playerId: String, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        let player = getPlayer(playerId)
        player?.enterFullscreen()
    }

    public func setMixWithOthers(_ playerId: String, mixWithOthers: Bool, completion: @escaping (FlutterError?) -> Void) {
        var player = getPlayer(playerId)
        player?.mixWithOthers = mixWithOthers
        updateAudioSession()
    }

    public func fetchMediaInfo(_ urlString: String, mimeType: String?) async -> (MediaInfo?, FlutterError?) {
        return await returnFlutterResult {
            guard let url = URL(string: urlString) else {
                throw BccmPlayerError.runtimeError("Invalid url")
            }
            let tracks = try await MediaInfoFetcher.fetchInfo(for: url, mimeType: mimeType)
            return tracks
        }
    }

    func updateAudioSession() {
        try? AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
        try? AVAudioSession.sharedInstance().setActive(true)
    }

}

public extension PlaybackApiImpl {
    func unclaimIfPrimary(_ playerId: String) {
        if primaryPlayerId != playerId { return }
        if let previousPrimaryPlayerId, let previous = getPlayer(previousPrimaryPlayerId) {
            setPrimary(previous.id, completion: { _ in })
            return
        }
        if let player = players.first(where: { $0 is AVQueuePlayerController }) {
            setPrimary(player.id, completion: { _ in })
            return
        }
        primaryPlayerId = nil
        playbackListener.onPrimaryPlayerChanged(PrimaryPlayerChangedEvent.make(withPlayerId: nil), completion: { _ in })
        assertionFailure("unclaimIfPrimary was called, but no player was given primary.")
    }
}
